blockchain/chain/src/store/chain_store.rs

```diff
@@ -646,6 +663,102 @@ where
     Ok(out)
 }
 
+#[cfg(feature = "json")]
+pub mod headchange_json {
+    use super::*;
+    use blocks::tipset_json::TipsetJsonRef;
+    use serde::Serialize;
+
+    #[derive(Serialize)]
+    #[serde(rename_all = "lowercase")]
+    #[serde(tag = "type", content = "val")]
+    pub enum HeadChangeJson<'a> {
+        Current(TipsetJsonRef<'a>),
+        Apply(TipsetJsonRef<'a>),
+        Revert(TipsetJsonRef<'a>),
+    }
+
+    #[derive(Debug, Clone)]
+    pub struct IndexToHeadChangeJson(pub usize, pub HeadChange);
+    impl<'a> From<&'a HeadChange> for HeadChangeJson<'a> {
+        fn from(wrapper: &'a HeadChange) -> Self {
+            match wrapper {
+                HeadChange::Current(tipset) => HeadChangeJson::Current(TipsetJsonRef(&tipset)),
+                HeadChange::Apply(tipset) => HeadChangeJson::Apply(TipsetJsonRef(&tipset)),
+                HeadChange::Revert(tipset) => HeadChangeJson::Revert(TipsetJsonRef(&tipset)),
+            }
+        }
+    }
+
+    #[derive(Clone)]
+    pub enum EventsPayload {
+        TaskCancel(usize, ()),
+        SubHeadChanges(IndexToHeadChangeJson),
+    }
+
+    impl EventsPayload {
+        pub fn sub_head_changes(&self) -> Option<&IndexToHeadChangeJson> {
+            match self {
+                EventsPayload::SubHeadChanges(s) => Some(s),
+                _ => None,
+            }
+        }
+
+        pub fn task_cancel(&self) -> Option<(usize, ())> {
+            match self {
+                EventsPayload::TaskCancel(val, _) => Some((*val, ())),
+                _ => None,
+            }
+        }
```

Comment:
These make no sense to me, why would you ever need this that you wouldn't be able to do with a match. (I see where it's used, but it doesn't seem clear to me why

Reply:
This is for the purpose of using it in filter closures. I want to be able to say task_cancel().map(|s|//operation goes here), I find it match cleaner than adding a match in a closure